libraries/AP_HAL_ChibiOS/Semaphores.cpp

```diff
@@ -78,4 +80,57 @@ void Semaphore::assert_owner(void)
     osalDbgAssert(check_owner(), "owner");
 }
 
+#if CH_CFG_USE_SEMAPHORES == TRUE
+BinarySemaphore::BinarySemaphore(bool initial_state) :
+    AP_HAL::BinarySemaphore(initial_state)
+{
+    static_assert(sizeof(_lock) >= sizeof(semaphore_t), "invalid semaphore_t size");
+    auto *sem = (binary_semaphore_t *)_lock;
+    /*
+      the initial_state in ChibiOS binary semaphores is 'taken', so we
+      need to negate it for the ArduPilot semantics
+     */
+    chBSemObjectInit(sem, !initial_state);
+}
+
+bool BinarySemaphore::wait(uint32_t timeout_us)
+{
+    auto *sem = (binary_semaphore_t *)_lock;
+    if (timeout_us == 0) {
+        return chBSemWaitTimeout(sem, TIME_IMMEDIATE) == MSG_OK;
+    }
+    // loop waiting for 60ms at a time. This ensures we can wait for
+    // any amount of time even on systems with a 16 bit timer
+    while (timeout_us > 0) {
+        const uint32_t us = MIN(timeout_us, 60000U);
+        if (chBSemWaitTimeout(sem, TIME_US2I(us)) == MSG_OK) {
+            return true;
+        }
+        timeout_us -= us;
+    }
+    return false;
+}
+
+bool BinarySemaphore::wait_blocking(void)
+{
+    auto *sem = (binary_semaphore_t *)_lock;
+    return chBSemWait(sem) == MSG_OK;
+}
+
+void BinarySemaphore::signal(void)
+{
+    auto *sem = (binary_semaphore_t *)_lock;
+    chBSemSignal(sem);
+}
+
+void BinarySemaphore::signal_ISR(void)
```

Comment:
I think you need the "I" version of this as well as this version in order to be useful. Usually you know if you are in an ISR and typically will already have a lock.

Reply:
none of the use cases so far have the lock. It is typically called as the last thing before you return from the interrupt.
The reason I'm reluctant to have an I version is that is a ChibiOS specific concept, and this is a HAL level call, uniform on SITL, Linux, ESP32 and ChibiOS. If we run across a case where we really need it (which would have to be in a ChibiOS specific driver) then we can address it, or the caller could unlock before making this call

Reply:
if/when we do have a use case. I think we'd do this:
```
void BinarySemaphore::signal_ISR(void)
{
    auto *sem = (binary_semaphore_t *)_lock;
    const bool was_locked = !port_irq_enabled(port_get_irq_status());
    if (!was_locked) {
      chSysLockFromISR();
    }
    chBSemSignalI(sem);
    if (!was_locked) {
      chSysUnlockFromISR();
   }
}
```
that is what we do in hrt.c for hrt_micros64()